flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSink.java

```diff
@@ -0,0 +1,273 @@
+	private final String projectName;
+	private final String topicName;
+	private final String hostAndPortForEmulator;
+
+	private transient Publisher publisher;
+
+	private PubSubSink(
+		Credentials credentials,
+		SerializationSchema<IN> serializationSchema,
+		String projectName,
+		String topicName,
+		String hostAndPortForEmulator) {
+		this.failureHandler = new FailureHandler();
+		this.credentials = credentials;
+		this.serializationSchema = serializationSchema;
+		this.projectName = projectName;
+		this.topicName = topicName;
+		this.hostAndPortForEmulator = hostAndPortForEmulator;
+	}
+
+	private transient ManagedChannel managedChannel = null;
+	private transient TransportChannel channel = null;
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		Publisher.Builder builder = Publisher
+			.newBuilder(ProjectTopicName.of(projectName, topicName))
+			.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+
+		if (hostAndPortForEmulator != null) {
+			managedChannel = ManagedChannelBuilder
+				.forTarget(hostAndPortForEmulator)
+				.usePlaintext(true) // This is 'Ok' because this is ONLY used for testing.
+				.build();
+			channel = GrpcTransportChannel.newBuilder().setManagedChannel(managedChannel).build();
+			builder.setChannelProvider(FixedTransportChannelProvider.create(channel))
+					.setCredentialsProvider(NoCredentialsProvider.create());
+		}
+
+		publisher = builder.build();
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		shutdownPublisher();
+		shutdownTransportChannel();
+		shutdownManagedChannel();
+	}
+
+	private void shutdownPublisher() {
+		try {
+			if (publisher != null) {
+				publisher.shutdown();
+			}
+		} catch (Exception e) {
+			LOG.info("Shutting down Publisher failed.", e);
+		}
+	}
+
+	private void shutdownTransportChannel() {
+		if (channel == null) {
+			return;
+		}
+		try {
+			channel.close();
+		} catch (Exception e) {
+			LOG.info("Shutting down TransportChannel failed.", e);
+		}
+	}
+
+	private void shutdownManagedChannel() {
+		if (managedChannel == null) {
+			return;
+		}
+		try {
+			managedChannel.shutdownNow();
+			managedChannel.awaitTermination(1000L, TimeUnit.MILLISECONDS);
+		} catch (Exception e) {
+			LOG.info("Shutting down ManagedChannel failed.", e);
+		}
+	}
+
+	@Override
+	public void invoke(IN message, SinkFunction.Context context) {
+		PubsubMessage pubsubMessage = PubsubMessage
+			.newBuilder()
+			.setData(ByteString.copyFrom(serializationSchema.serialize(message)))
+			.build();
+		ApiFutures.addCallback(publisher.publish(pubsubMessage), failureHandler, directExecutor());
+	}
+
+	/**
+	 * Create a builder for a new PubSubSink.
+	 *
+	 * @param <IN> The generic of the type that is to be written into the sink.
+	 * @return a new PubSubSinkBuilder instance
+	 */
+	public static <IN> PubSubSinkBuilder<IN> newBuilder(SerializationSchema<IN> serializationSchema, String projectName, String topicName) {
```

Comment:
Previously it was like that and this gave several issues:

* To use it would require awkward syntax. It would be `PubSubSource.<SomeType>newBuilder().withEtc())`
* Another is that these arguments are the only arguments that are required. And I kinda like this pattern because it makes it impossible to forget to add them.

Reply:
Nice I didn't know about this pattern, I love the step builder pattern, thank you for pointing that one out :) Added it to both the source and sink